File: shared/db/pagination.ts
/**
 * Centralized pagination helper for list endpoints.
 *
 * Optimization rules:
 *  - Pagination is opt-in. If the caller does NOT pass `page` or `limit`, the
 *    current behavior is preserved (returning a plain array) so existing UIs
 *    that consume `Array<T>` keep working.
 *  - When the caller passes `page` or `limit`, the service may return either:
 *      * `{ items, total, page, limit, pages }`  (recommended for new UIs), OR
 *      * a plain array sliced to the page (when `paginate` is "soft").
 *
 * Both shapes are wrapped in `ServiceResult<T>` upstream, so the API envelope
 * is unchanged.
 */
export type PaginationInput = {
    page?: string | number;
    limit?: string | number;
};

export type PaginationParams = {
    enabled: boolean;
    page: number;
    limit: number;
    skip: number;
};

export type Paginated<T> = {
    items: T[];
    total: number;
    page: number;
    limit: number;
    pages: number;
};

const DEFAULT_LIMIT = 25;
const MAX_LIMIT = 200;

export function parsePagination(input: PaginationInput | undefined, defaults: { defaultLimit?: number; maxLimit?: number } = {}): PaginationParams {
    const defaultLimit = defaults.defaultLimit ?? DEFAULT_LIMIT;
    const maxLimit = defaults.maxLimit ?? MAX_LIMIT;

    const hasPage = input?.page !== undefined && input.page !== "" && input.page !== null;
    const hasLimit = input?.limit !== undefined && input.limit !== "" && input.limit !== null;
    const enabled = hasPage || hasLimit;

    let page = Number(input?.page ?? 1);
    let limit = Number(input?.limit ?? defaultLimit);
    if (!Number.isFinite(page) || page < 1) page = 1;
    if (!Number.isFinite(limit) || limit < 1) limit = defaultLimit;
    if (limit > maxLimit) limit = maxLimit;

    return {
        enabled,
        page,
        limit,
        skip: (page - 1) * limit
    };
}

export function buildPaginatedResponse<T>(items: T[], total: number, params: PaginationParams): Paginated<T> {
    return {
        items,
        total,
        page: params.page,
        limit: params.limit,
        pages: Math.max(1, Math.ceil(total / params.limit))
    };
}

File: backend-go/internal/api/pagination.go
package api

import (
	"net/url"
	"strconv"
)

// PaginationParams mirrors the helper from old-app/shared/db/pagination.ts.
type PaginationParams struct {
	Enabled bool
	Page    int
	Limit   int
	Skip    int
}

// Paginated mirrors the original Paginated<T> shape exactly so the React
// frontend's `usePaginatedList` consumes the same fields.
type Paginated[T any] struct {
	Items []T `json:"items"`
	Total int `json:"total"`
	Page  int `json:"page"`
	Limit int `json:"limit"`
	Pages int `json:"pages"`
}

// ParsePagination reads `page` and `limit` from a URL query, applying the
// same defaults as parsePagination in the Node helper.
func ParsePagination(q url.Values) PaginationParams {
	const defaultLimit = 25
	const maxLimit = 200

	hasPage := q.Get("page") != ""
	hasLimit := q.Get("limit") != ""
	enabled := hasPage || hasLimit

	page, _ := strconv.Atoi(q.Get("page"))
	if page < 1 {
		page = 1
	}
	limit, _ := strconv.Atoi(q.Get("limit"))
	if limit < 1 {
		limit = defaultLimit
	}
	if limit > maxLimit {
		limit = maxLimit
	}

	return PaginationParams{
		Enabled: enabled,
		Page:    page,
		Limit:   limit,
		Skip:    (page - 1) * limit,
	}
}

// BuildPaginated mirrors `buildPaginatedResponse`.
func BuildPaginated[T any](items []T, total int, p PaginationParams) Paginated[T] {
	pages := total / p.Limit
	if total%p.Limit != 0 {
		pages++
	}
	if pages < 1 {
		pages = 1
	}
	return Paginated[T]{
		Items: items,
		Total: total,
		Page:  p.Page,
		Limit: p.Limit,
		Pages: pages,
	}
}
